### PR TITLE
Kubectl get deployments

### PR DESCRIPTION
**Pull request by janetkuo**
Print different `replicas` of deployments. 

``` console
$ kubectl get deployments nginx-deployment
NAME               DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
nginx-deployment   3         3         3            3           3m
```

cc @nikhiljindal @kargakis @bgrant0607 @kubernetes/kubectl

### CODE REVIEW

**Comment by kargakis** [comment]
LGTM

General observation: UPDATEDREPLICAS is probably confusing

**Comment by bgrant0607** [comment]
Please use `_` as the word separator (and add this to kubectl conventions if it isn't there).

How about REPLICAS (desired), CURRENT, UP_TO_DATE, and AVAILABLE.

At some point, we should reconcile with RC and RS columns.

**Comment by bgrant0607** [comment]
RC columns are:
```
"CONTROLLER", "CONTAINER(S)", "IMAGE(S)", "SELECTOR", "REPLICAS", "AGE"
```

Fields that contain lists (containers, images, selector) make the output wide and hard to read, so I don't think we should just copy this. For instance, I don't think containers and selector are very important here. Existing label options should satisfy most of the use cases for the selector, and images (and their tags) are more important than the container names.

I'd list the first image, appended with +N for the number of other containers, if any.

**Comment by bgrant0607** [comment]
See also #19775

**Comment by janetkuo** [comment]
If we use IMAGE(S), and REPLICAS, CURRENT, UP_TO_DATE, and AVAILABLE, it looks like this:
```console
$ kubectl get deployments nginx-deployment
NAME               IMAGE(S)   REPLICAS  CURRENT  UP_TO_DATE  AVAILABLE   AGE
nginx-deployment   nginx(+2)  3         3        3           3           1s
```

I think CURRENT, UP_TO_DATE, and AVAILABLE are a bit vague without "REPLICAS", but REPLICAS is too long. hpa uses PODS instead of REPLICAS in printer header. So how about:
```console
$ kubectl get deployments nginx-deployment
NAME               IMAGE(S)   DESIRED_PODS  CURRENT_PODS  UP_TO_DATE_PODS  AVAILABLE_PODS   AGE
nginx-deployment   nginx(+2)  3             3             3                3                1s
```
or we can use two-line header:
```console
$ kubectl get deployments nginx-deployment
NAME               IMAGE(S)   DESIRED   CURRENT  UP_TO_DATE  AVAILABLE   AGE
                              REPLICAS  REPLICAS REPLICAS    REPLICAS 
nginx-deployment   nginx(+2)  3         3        3           3           1s
```
or we merge replicas together:
```console
$ kubectl get deployments nginx-deployment
NAME               IMAGE(S)   CURRENT/UP_TO_DATE/AVAILABLE/DESIRED_PODS   AGE
nginx-deployment   nginx(+2)  3/3/3/3                                     1s
```

**Comment by ghodss** [comment]
@janetkuo I would go with your first option, but without IMAGES.

While its true that the columns are vague at first, I think it's just a small learning curve that's okay to incur, especially if `kubectl describe` writes out the equivalent columns with the "repliacs" suffix. So:

```console
$ kubectl get deployments nginx-deployment
NAME              DESIRED   CURRENT  UP_TO_DATE  AVAILABLE   AGE
nginx-deployment  3         3        3           3           1s

$ kubectl describe deployments nginx-deployment
Name: nginx-deployment
Desired replicas: 3
Current replicas: 3
Up to date replicas: 3
Available replicas: 3
...
```

That way if someone doesn't understand the column names they can just run describe and pretty quickly realize what they mean. We also can update the deployment user guide to explain the columns.

As for listing IMAGE(S):
# It can get very wide for images in a local registry + a customized version tag (one sample image name from us: `docker-registry-vip.dev.box.net/jenkins/convert-mon:0.0.0-21.2015_12_31_22_37_37.56e95-image-16`)
# It provides minimal value here since you might accidentally show a sidecar image
# The name is probably enough to have an idea of what images exist.

I wrote out a similar case in #19775.

**Comment by bgrant0607** [comment]
I like the proposal from @ghodss. 

We can add images later, maybe under `-o wide`. FWIW, I intended the +N for other containers in the pod, not other replicas or revisions. We could strip off the registry. It might even make sense to just show the tag.

**Comment by bgrant0607** [comment]
Note that status is a long-requested feature #7483. We should add status summaries to ReplicaSet as well at some point.

**Comment by ghodss** [comment]
One other suggestion - can we do `UP TO DATE` with spaces instead of underscores? I think underscores looks very unnatural, and the padding between columns is two spaces so there's minimal risk of it seeming like it's multiple columns. See docker's ps command as an example of using spaces:

```console
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
$
```

I personally think it would look much worse with underscores:

```console
$ docker ps
CONTAINER_ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
$
```

netstat uses spaces:

```console
$ netstat -anlp
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:2049            0.0.0.0:*               LISTEN      -
tcp        0      0 0.0.0.0:33895           0.0.0.0:*               LISTEN      -
...
```

Other examples usually make the columns more cryptic rather than resort to word separators:

```console
$ ps axu
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.0  0.3  46264  6156 ?        Ss   Jan24   0:02 /usr/lib/systemd/systemd --switched-root --system --deserialize 22
root         2  0.0  0.0      0     0 ?        S    Jan24   0:00 [kthreadd]
root         3  0.0  0.0      0     0 ?        S    Jan24   0:16 [ksoftirqd/0]
...
$ fleetctl list-unit-files
UNIT          HASH    DSTATE   STATE    TMACHINE
hello.service e55c0ae inactive inactive -
$ w
 1:44  up 11 days,  7:41, 3 users, load averages: 2.08 1.92 2.13
USER     TTY      FROM              LOGIN@  IDLE WHAT
sam      console  -                13Jan16 11days -
sam      s000     -                13Jan16 10:51 -bash
sam      s001     -                13Jan16     - tmux
```

I haven't found any command line tool that uses underscores as separators.

tl;dr I think spaces are the best balance between readability and usability.

**Comment by bgrant0607** [comment]
@ghodss The reason I don't like spaces in column headers or contents is because that's unfriendly to column extraction (e.g., `cut`).

**Comment by bgrant0607** [comment]
I also find the netstat output hard to read due to the single spaces between some columns (esp. Send-Q and "Local Address").

**Comment by bgrant0607** [comment]
`-o wide` (#19775) can be implemented in a followup PR and doesn't necessarily need to be in 1.2. Just omit those wide fields from this PR.

**Comment by ghodss** [comment]
> @ghodss The reason I don't like spaces in column headers or contents is because that's unfriendly to column extraction (e.g., cut).

If you're extracting columns, there's a high likelihood you're doing some processing and would want to use --no-headers anyways. I would suggest starting with the friendlier version (single spaces) as convention and moving away if there are complaints.

Also as I was thinking more about this case, up-to-date is often hyphenated in plan English, so maybe UP-TO-DATE is acceptable.

**Comment by janetkuo** [comment]
@bgrant0607 Comments addressed, PTAL. 
Sample output:
```console 
$ kubectl get deployments nginx-deployment
NAME               DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
nginx-deployment   3         3         3            3           3m
```
kubectl convention doc is updated too.

File: docs/devel/kubectl-conventions.md [hunk 1]
```diff
@@ -110,7 +110,7 @@ Updated: 8/27/2015
   * However, affordances are made for simple parsing of `get` output
 * Only errors should be directed to stderr
 * `get` commands should output one row per resource, and one resource per row
-  * Column titles and values should not contain spaces in order to facilitate commands that break lines into fields: cut, awk, etc.
+  * Column titles and values should not contain spaces in order to facilitate commands that break lines into fields: cut, awk, etc. Instead, use `_` as the word separator.
```
Comment (bgrant0607) [inline docs/devel/kubectl-conventions.md]: You use dash. :-)

Could you change this to dash and then also change CLUSTER_IP and EXTERNAL_IP to match the convention?

**Comment by bgrant0607** [comment]
Mostly LG except for underscore vs. dash.

File: docs/devel/kubectl-conventions.md (hunk 1 above)
Comment (janetkuo) [inline docs/devel/kubectl-conventions.md]: Done

**Comment by janetkuo** [comment]
@bgrant0607 All comments addressed. PTAL.

**Comment by bgrant0607** [comment]
LGTM
